crates/subspace-farmer/src/utils/segment_reconstruction.rs

```diff
@@ -0,0 +1,79 @@
+use std::iter;
+use futures::stream::{FuturesOrdered, FuturesUnordered};
+use futures::StreamExt;
+use subspace_core_primitives::{Piece, PieceIndex, SegmentIndex, PIECES_IN_SEGMENT, RECORD_SIZE};
+use subspace_farmer_components::plotting::PieceGetter;
+use tracing::{error, info};
+use subspace_archiving::reconstructor::Reconstructor;
+
+const SEGMENT_SIZE: u32 = RECORD_SIZE * PIECES_IN_SEGMENT / 2;
+
+pub struct SegmentReconstruction<PG> {
+    piece_getter: PG,
+}
+
+impl<PG: PieceGetter> SegmentReconstruction<PG> {
+    pub fn new(piece_getter: PG) -> Self {
+        Self { piece_getter }
+    }
+
+    pub async fn get_missing_piece(&self, missing_piece_index: PieceIndex) -> Option<Piece> {
```

Comment:
It doesn't just get it, it reconstructs or recovers, so the name of the methods/function should reflect that for better readability. If we get/retrieve everything it is harder to read code.

This one looks like a function to me as well. Especially because in order to get pieces you'll already have to have `piece_getter` in the same context. This is not Java, it doesn't have to be a class for everything, functions are fine too.